File: SJF/sjf.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iomanip>
#include <string>

struct Process {
    std::string pid;
    int arrivalTime;
    int burstTime;
    int completionTime;
    int turnaroundTime;
    int waitingTime;
    int responseTime;
    bool completed;
    int originalIndex;
};

// Function to calculate Shortest Job First (Non-preemptive)
void calculateSJF(std::vector<Process>& processes, std::vector<Process>& timeline) {
    int n = processes.size();
    if (n == 0) return;

    // Initialize completion state
    for (auto& p : processes) {
        p.completed = false;
    }

    int currentTime = 0;
    int completedCount = 0;

    while (completedCount < n) {
        int selectedIdx = -1;
        int minBurstTime = 1e9;
        int minArrivalTime = 1e9;

        // Find the process that is ready and has the shortest burst time
        for (int i = 0; i < n; i++) {
            if (!processes[i].completed && processes[i].arrivalTime <= currentTime) {
                if (processes[i].burstTime < minBurstTime) {
                    minBurstTime = processes[i].burstTime;
                    minArrivalTime = processes[i].arrivalTime;
                    selectedIdx = i;
                } else if (processes[i].burstTime == minBurstTime) {
                    // Tie-breaker: choose process with earlier arrival time
                    if (processes[i].arrivalTime < minArrivalTime) {
                        minArrivalTime = processes[i].arrivalTime;
                        selectedIdx = i;
                    }
                }
            }
        }

        if (selectedIdx == -1) {
            // CPU is idle. Advance to the next arriving process.
            int nextArrival = 1e9;
            for (int i = 0; i < n; i++) {
                if (!processes[i].completed && processes[i].arrivalTime < nextArrival) {
                    nextArrival = processes[i].arrivalTime;
                }
            }
            // Add Idle block to timeline
            Process idle;
            idle.pid = "Idle";
            idle.arrivalTime = currentTime;
            idle.burstTime = nextArrival - currentTime;
            idle.completionTime = nextArrival;
            timeline.push_back(idle);

            currentTime = nextArrival;
        } else {
            processes[selectedIdx].completionTime = currentTime + processes[selectedIdx].burstTime;
            processes[selectedIdx].turnaroundTime = processes[selectedIdx].completionTime - processes[selectedIdx].arrivalTime;
            processes[selectedIdx].waitingTime = processes[selectedIdx].turnaroundTime - processes[selectedIdx].burstTime;
            processes[selectedIdx].responseTime = currentTime - processes[selectedIdx].arrivalTime;
            processes[selectedIdx].completed = true;

            timeline.push_back(processes[selectedIdx]);
            currentTime = processes[selectedIdx].completionTime;
            completedCount++;
        }
    }
}

int main(int argc, char* argv[]) {
    bool jsonMode = false;
    if (argc > 1 && std::string(argv[1]) == "--json") {
        jsonMode = true;
    }

    if (jsonMode) {
        int n;
        if (!(std::cin >> n) || n <= 0) {
            std::cout << "{\"error\": \"Invalid process count\"}\n";
            return 1;
        }

        std::vector<Process> processes(n);
        for (int i = 0; i < n; i++) {
            std::cin >> processes[i].pid >> processes[i].arrivalTime >> processes[i].burstTime;
            processes[i].originalIndex = i;
        }

        std::vector<Process> timeline;
        calculateSJF(processes, timeline);

        // Print JSON Output
        std::cout << "{\n  \"processes\": [\n";
        for (int i = 0; i < n; i++) {
            std::cout << "    {\n"
                      << "      \"pid\": \"" << processes[i].pid << "\",\n"
                      << "      \"arrivalTime\": " << processes[i].arrivalTime << ",\n"
                      << "      \"burstTime\": " << processes[i].burstTime << ",\n"
                      << "      \"completionTime\": " << processes[i].completionTime << ",\n"
                      << "      \"turnaroundTime\": " << processes[i].turnaroundTime << ",\n"
                      << "      \"waitingTime\": " << processes[i].waitingTime << ",\n"
                      << "      \"responseTime\": " << processes[i].responseTime << "\n"
                      << "    }" << (i == n - 1 ? "" : ",") << "\n";
        }
        std::cout << "  ],\n  \"timeline\": [\n";
        for (size_t i = 0; i < timeline.size(); i++) {
            std::cout << "    {\n"
                      << "      \"pid\": \"" << timeline[i].pid << "\",\n"
                      << "      \"startTime\": " << (timeline[i].completionTime - timeline[i].burstTime) << ",\n"
                      << "      \"endTime\": " << timeline[i].completionTime << ",\n"
                      << "      \"isIdle\": " << (timeline[i].pid == "Idle" ? "true" : "false") << "\n"
                      << "    }" << (i == timeline.size() - 1 ? "" : ",") << "\n";
        }
        std::cout << "  ]\n}\n";

    } else {
        int n;
        std::cout << "=== SJF CPU Scheduling Simulator ===\n";
        std::cout << "Enter number of processes: ";
        if (!(std::cin >> n) || n <= 0) {
            std::cerr << "Invalid number of processes.\n";
            return 1;
        }

        std::vector<Process> processes(n);
        for (int i = 0; i < n; i++) {
            std::cout << "\nProcess " << i + 1 << " ID (e.g. P1): ";
            std::cin >> processes[i].pid;
            std::cout << "Arrival Time: ";
            std::cin >> processes[i].arrivalTime;
            std::cout << "Burst Time: ";
            std::cin >> processes[i].burstTime;
            processes[i].originalIndex = i;
        }

        std::vector<Process> timeline;
        calculateSJF(processes, timeline);

        // Display Gantt Chart
        std::cout << "\n=== Gantt Chart ===\n";
        for (const auto& block : timeline) {
            int startTime = block.completionTime - block.burstTime;
            std::cout << "[" << block.pid << " " << startTime << "-" << block.completionTime << "] ";
        }
        std::cout << "\n\n";

        // Display Metrics Table
        std::cout << std::left << std::setw(12) << "Process ID" 
                  << std::setw(15) << "Arrival Time" 
                  << std::setw(12) << "Burst Time" 
                  << std::setw(18) << "Completion Time" 
                  << std::setw(18) << "Turnaround Time" 
                  << std::setw(15) << "Waiting Time" 
                  << std::setw(15) << "Response Time" << "\n";
                  
        double totalTAT = 0, totalWT = 0, totalRT = 0;
        for (const auto& p : processes) {
            totalTAT += p.turnaroundTime;
            totalWT += p.waitingTime;
            totalRT += p.responseTime;
            
            std::cout << std::left << std::setw(12) << p.pid 
                      << std::setw(15) << p.arrivalTime 
                      << std::setw(12) << p.burstTime 
                      << std::setw(18) << p.completionTime 
                      << std::setw(18) << p.turnaroundTime 
                      << std::setw(15) << p.waitingTime 
                      << std::setw(15) << p.responseTime << "\n";
        }

        std::cout << std::fixed << std::setprecision(2);
        std::cout << "\nAverage Turnaround Time: " << (totalTAT / n) << "\n";
        std::cout << "Average Waiting Time: " << (totalWT / n) << "\n";
        std::cout << "Average Response Time: " << (totalRT / n) << "\n";
    }

    return 0;
}

File: SJF/SJF.js
const processes = [];
const colors = [
    '#3b82f6', // blue
    '#8b5cf6', // purple
    '#10b981', // emerald
    '#f59e0b', // amber
    '#ec4899', // pink
    '#14b8a6', // teal
    '#f43f5e', // rose
    '#6366f1'  // indigo
];



document.getElementById('add-process-btn').addEventListener('click', addProcess);
document.getElementById('load-sample-btn').addEventListener('click', loadSampleScenario);
document.getElementById('clear-all-btn').addEventListener('click', clearAll);
document.getElementById('simulate-btn').addEventListener('click', runSimulation);

function updateProcessTable() {
    const tbody = document.getElementById('process-list-body');
    tbody.innerHTML = '';

    if (processes.length === 0) {
        tbody.innerHTML = '<tr><td colspan="4" style="text-align: center; color: var(--text-secondary);">No processes added yet. Click "Load Sample Scenario" or enter details above.</td></tr>';
        return;
    }

    processes.forEach((p, idx) => {
        const tr = document.createElement('tr');
        tr.innerHTML = `
                    <td><strong>${p.pid}</strong></td>
                    <td>${p.arrivalTime}</td>
                    <td>${p.burstTime}</td>
                    <td><button class="btn btn-danger" style="padding: 0.3rem 0.6rem; font-size: 0.8rem;" onclick="removeProcess(${idx})">Remove</button></td>
                `;
        tbody.appendChild(tr);
    });
}

function addProcess() {
    const pidInput = document.getElementById('pid');
    const atInput = document.getElementById('arrival-time');
    const btInput = document.getElementById('burst-time');

    let pid = pidInput.value.trim();
    const arrivalTime = parseInt(atInput.value);
    const burstTime = parseInt(btInput.value);

    if (isNaN(arrivalTime) || arrivalTime < 0) {
        alert('Arrival Time must be a non-negative integer.');
        return;
    }
    if (isNaN(burstTime) || burstTime <= 0) {
        alert('Burst Time must be a positive integer.');
        return;
    }

    if (!pid) {
        pid = 'P' + (processes.length + 1);
    }

    if (processes.some(p => p.pid === pid)) {
        alert('Process ID must be unique.');
        return;
    }

    processes.push({ pid, arrivalTime, burstTime });
    updateProcessTable();

    pidInput.value = '';
    atInput.value = '0';
    btInput.value = '5';
}

function removeProcess(idx) {
    processes.splice(idx, 1);
    updateProcessTable();
}

function loadSampleScenario() {
    processes.length = 0;
    processes.push({ pid: 'P1', arrivalTime: 0, burstTime: 5 });
    processes.push({ pid: 'P2', arrivalTime: 1, burstTime: 3 });
    processes.push({ pid: 'P3', arrivalTime: 2, burstTime: 8 });
    processes.push({ pid: 'P4', arrivalTime: 3, burstTime: 6 });
    updateProcessTable();
}

function clearAll() {
    processes.length = 0;
    updateProcessTable();
    document.getElementById('gantt-section').style.display = 'none';
    document.getElementById('results-section').style.display = 'none';
}

window.removeProcess = removeProcess;

// JS Fallback SJF Scheduling logic
function simulateSJF_JS(procList) {
    const localProcesses = JSON.parse(JSON.stringify(procList)).map((p, idx) => ({
        ...p,
        originalIndex: idx,
        completed: false
    }));

    const timeline = [];
    let currentTime = 0;
    let completedCount = 0;
    const simulatedProcesses = [];

    while (completedCount < localProcesses.length) {
        const ready = localProcesses.filter(p => !p.completed && p.arrivalTime <= currentTime);

        if (ready.length === 0) {
            const uncompleted = localProcesses.filter(p => !p.completed);
            const nextArrival = Math.min(...uncompleted.map(p => p.arrivalTime));

            timeline.push({
                pid: 'Idle',
                startTime: currentTime,
                endTime: nextArrival,
                isIdle: true
            });
            currentTime = nextArrival;
            continue;
        }

        ready.sort((a, b) => {
            if (a.burstTime !== b.burstTime) {
                return a.burstTime - b.burstTime;
            }
            return a.arrivalTime - b.arrivalTime;
        });

        const selected = ready[0];
        const startTime = currentTime;
        const endTime = currentTime + selected.burstTime;

        selected.startTime = startTime;
        selected.completionTime = endTime;
        selected.turnaroundTime = selected.completionTime - selected.arrivalTime;
        selected.waitingTime = selected.turnaroundTime - selected.burstTime;
        selected.responseTime = selected.startTime - selected.arrivalTime;
        selected.completed = true;

        timeline.push({
            pid: selected.pid,
            startTime: startTime,
            endTime: endTime,
            color: colors[selected.originalIndex % colors.length],
            isIdle: false
        });

        simulatedProcesses.push(selected);
        currentTime = endTime;
        completedCount++;
    }

    return { processes: simulatedProcesses, timeline };
}

function runSimulation() {
    if (processes.length === 0) {
        alert('Please add at least one process to simulate.');
        return;
    }

    const resultData = simulateSJF_JS(processes);
    renderSimulation(resultData.processes, resultData.timeline);
}

function renderSimulation(simulatedProcesses, timeline) {
    const ganttTimeline = document.getElementById('gantt-timeline');
    const ganttTimeLabels = document.getElementById('gantt-time-labels');
    ganttTimeline.innerHTML = '';
    ganttTimeLabels.innerHTML = '';

    const totalSimTime = timeline[timeline.length - 1].endTime;

    timeline.forEach((block, idx) => {
        const blockDuration = block.endTime - block.startTime;
        const percentage = (blockDuration / totalSimTime) * 100;

        const blockEl = document.createElement('div');
        blockEl.className = 'gantt-block' + (block.isIdle ? ' idle' : '');
        blockEl.style.width = `${percentage}%`;
        if (!block.isIdle) {
            const processIndex = processes.findIndex(p => p.pid === block.pid);
            blockEl.style.backgroundColor = colors[processIndex !== -1 ? processIndex : idx % colors.length];
        }
        blockEl.innerHTML = `<div>${block.pid}</div><div style="font-size: 0.75rem; font-weight: normal; opacity: 0.8;">(${blockDuration})</div>`;
        ganttTimeline.appendChild(blockEl);
    });

    // Add time labels
    const zeroLabel = document.createElement('span');
    zeroLabel.className = 'gantt-time-label';
    zeroLabel.style.left = '0%';
    zeroLabel.textContent = '0';
    ganttTimeLabels.appendChild(zeroLabel);

    let cumulativePercent = 0;
    timeline.forEach((block) => {
        const blockDuration = block.endTime - block.startTime;
        cumulativePercent += (blockDuration / totalSimTime) * 100;

        const label = document.createElement('span');
        label.className = 'gantt-time-label';
        label.style.left = `${cumulativePercent}%`;
        label.textContent = block.endTime;
        ganttTimeLabels.appendChild(label);
    });

    document.getElementById('gantt-section').style.display = 'block';

    // Populate Metrics Table in original order
    const resultsTbody = document.getElementById('results-table-body');
    resultsTbody.innerHTML = '';

    let sumTat = 0;
    let sumWt = 0;
    let sumRt = 0;

    processes.forEach((p) => {
        const calculated = simulatedProcesses.find(lp => lp.pid === p.pid);
        if (calculated) {
            sumTat += calculated.turnaroundTime;
            sumWt += calculated.waitingTime;
            sumRt += calculated.responseTime;

            const tr = document.createElement('tr');
            tr.innerHTML = `
                        <td><strong>${calculated.pid}</strong></td>
                        <td>${calculated.arrivalTime}</td>
                        <td>${calculated.burstTime}</td>
                        <td>${calculated.completionTime}</td>
                        <td>${calculated.turnaroundTime}</td>
                        <td>${calculated.waitingTime}</td>
                        <td>${calculated.responseTime}</td>
                    `;
            resultsTbody.appendChild(tr);
        }
    });

    const avgTat = (sumTat / processes.length).toFixed(2);
    const avgWt = (sumWt / processes.length).toFixed(2);
    const avgRt = (sumRt / processes.length).toFixed(2);

    document.getElementById('avg-tat').textContent = avgTat;
    document.getElementById('avg-wt').textContent = avgWt;
    document.getElementById('avg-rt').textContent = avgRt;

    document.getElementById('results-section').style.display = 'block';
}

updateProcessTable();